File: src/nios/intelligence/truth/TruthLayer.ts
/**
 * Truth Layer — Phase 1
 * Every fact must be evidence-bound. Unsupported facts trigger research plan.
 */

import type { TruthRecord, VerificationReport } from "../contracts/types";

export interface TruthValidationResult {
  ok: boolean;
  truthRecords: TruthRecord[];
  unsupported: string[];
  researchPlan?: string[];
}

export function createTruthRecord(
  statement: string,
  source: string,
  evidence: string[],
  options?: {
    confidence?: number;
    jurisdiction?: string;
    knowledge_version?: string;
    verification_status?: TruthRecord["verification_status"];
  },
): TruthRecord {
  return {
    statement,
    evidence,
    source,
    confidence: options?.confidence ?? 1.0,
    timestamp: new Date().toISOString(),
    jurisdiction: options?.jurisdiction ?? "NP",
    knowledge_version: options?.knowledge_version,
    verification_status: options?.verification_status ?? "verified_deterministic",
  };
}

export function validateFacts(
  statements: Array<{ text: string; evidence: string[]; source: string }>,
): TruthValidationResult {
  const truthRecords: TruthRecord[] = [];
  const unsupported: string[] = [];

  for (const stmt of statements) {
    if (!stmt.evidence.length) {
      unsupported.push(stmt.text);
      continue;
    }
    truthRecords.push(
      createTruthRecord(stmt.text, stmt.source, stmt.evidence, {
        verification_status:
          stmt.source.startsWith("cap.engine") ? "verified_deterministic" : "verified_evidence",
      }),
    );
  }

  return {
    ok: unsupported.length === 0,
    truthRecords,
    unsupported,
    researchPlan:
      unsupported.length > 0
        ? [
            "cap.knowledge.nepal.search",
            "federation.web.allowlisted",
            "agent.researcher",
          ]
        : undefined,
  };
}

export function mergeVerificationWithTruth(
  report: VerificationReport,
  validation: TruthValidationResult,
): VerificationReport {
  return {
    ...report,
    ok: report.ok && validation.ok,
    truthRecords: [...report.truthRecords, ...validation.truthRecords],
    checks: [
      ...report.checks,
      {
        name: "truth_layer",
        passed: validation.ok,
        detail:
          validation.unsupported.length > 0
            ? `Unsupported: ${validation.unsupported.join("; ")}`
            : "All facts evidenced",
      },
    ],
  };
}

File: erp_bot/src/nios/intelligence/truth_layer.py
"""Truth Layer — Phase 1."""

from __future__ import annotations

from dataclasses import dataclass, field

from ..contracts.intelligence_contract import TruthRecord, utc_now


@dataclass
class TruthValidationResult:
    ok: bool
    truth_records: list[TruthRecord] = field(default_factory=list)
    unsupported: list[str] = field(default_factory=list)
    research_plan: list[str] | None = None


def create_truth_record(
    statement: str,
    source: str,
    evidence: list[str],
    *,
    confidence: float = 1.0,
    jurisdiction: str = "NP",
    verification_status: str = "verified_deterministic",
) -> TruthRecord:
    return TruthRecord(
        statement=statement,
        evidence=evidence,
        source=source,
        confidence=confidence,
        timestamp=utc_now(),
        jurisdiction=jurisdiction,
        verification_status=verification_status,
    )


def validate_facts(
    statements: list[dict[str, object]],
) -> TruthValidationResult:
    truth_records: list[TruthRecord] = []
    unsupported: list[str] = []

    for stmt in statements:
        text = str(stmt.get("text", ""))
        evidence = list(stmt.get("evidence") or [])
        source = str(stmt.get("source", "unknown"))
        if not evidence:
            unsupported.append(text)
            continue
        status = "verified_deterministic" if source.startswith("cap.engine") else "verified_evidence"
        truth_records.append(create_truth_record(text, source, evidence, verification_status=status))

    return TruthValidationResult(
        ok=len(unsupported) == 0,
        truth_records=truth_records,
        unsupported=unsupported,
        research_plan=(
            ["cap.knowledge.nepal.search", "federation.web.allowlisted", "agent.researcher"]
            if unsupported
            else None
        ),
    )
